### PR TITLE
Enable ACLMODE 

### PR DESCRIPTION
**Pull request by mattmacy**
This currently depends on https://github.com/zfsonlinux/zfs/pull/7728

However, if it looks like that will take a while we can also put ACLMODE under
a `__FreeBSD__` ifdef

Signed-off-by: Matt Macy <mmacy@FreeBSD.org>

### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:

- [ ] My code follows the ZFS on Linux [code style requirements](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [ ] I have updated the documentation accordingly.
- [ ] I have read the [**contributing** document](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md).
- [ ] I have added [tests](https://github.com/zfsonlinux/zfs/tree/master/tests) to cover my changes.
- [ ] All new and existing tests passed.
- [ ] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).

### CODE REVIEW

File: include/sys/fs/zfs.h [hunk 1]
```diff
@@ -115,7 +115,8 @@ typedef enum {
 	ZFS_PROP_READONLY,
 	ZFS_PROP_ZONED,
 	ZFS_PROP_SNAPDIR,
-	ZFS_PROP_PRIVATE,		/* not exposed to user, temporary */
+	ZFS_PROP_PRIVATE,		/* not exposed to the user, temporary */
+	ZFS_PROP_ACLMODE,
```
Comment (behlendorf) [inline include/sys/fs/zfs.h]: While this was the original placement, we should add this to the end to avoid reordering the enum.  There may be external consumer depending on it.

File: module/zcommon/zfs_prop.c [hunk 2]
```diff
@@ -333,6 +341,10 @@ zfs_prop_init(void)
 	zprop_register_index(ZFS_PROP_ACLTYPE, "acltype", ZFS_ACLTYPE_OFF,
 	    PROP_INHERIT, ZFS_TYPE_FILESYSTEM | ZFS_TYPE_SNAPSHOT,
 	    "noacl | posixacl", "ACLTYPE", acltype_table);
+	zprop_register_index(ZFS_PROP_ACLMODE, "aclmode", ZFS_ACL_DISCARD,
+	    PROP_INHERIT, ZFS_TYPE_FILESYSTEM,
+	    "discard | groupmask | passthrough | restricted", "ACLMODE",
+	    acl_mode_table);
```
Comment (behlendorf) [inline module/zcommon/zfs_prop.c]: I'd suggest we split the aclmode changes in to their own PR.

File: module/zcommon/zfs_prop.c [hunk 3]
```diff
@@ -379,8 +391,13 @@ zfs_prop_init(void)
 	zprop_register_index(ZFS_PROP_READONLY, "readonly", 0, PROP_INHERIT,
 	    ZFS_TYPE_FILESYSTEM | ZFS_TYPE_VOLUME, "on | off", "RDONLY",
 	    boolean_table);
+#ifdef __FreeBSD__
+	zprop_register_index(ZFS_PROP_ZONED, "jailed", 0, PROP_INHERIT,
+	    ZFS_TYPE_FILESYSTEM, "on | off", "JAILED", boolean_table);
+#else
```
Comment (behlendorf) [inline module/zcommon/zfs_prop.c]: Can this chunk be included latter with the rest of the changes for jails.  I expect there must be additional changes to the common code, at a minimum to the command line tools.

Comment (mattmacy) [reply module/zcommon/zfs_prop.c]: there are changes to `/cmd/zfs/zfs_main.c` The rest are in files added to the libzfs for FreeBSD

**Comment by behlendorf** [comment]
Closing.  This work will be done post-FreeBSD merge when it will be easier to integrate and verify the functionality.  See https://github.com/zfsonlinux/zfs/pull/7728#issuecomment-562710621 for details.
